File: gcd_string_high.py
import math

class Solution(object):
    def gcdOfStrings(self, str1, str2):
        len1 , len2= len(str1),len(str2)
        len_gcd = self.gcd(len1,len2)
        if len1 % len_gcd !=0 or len2 % len_gcd !=0 : return ""
        c =str1[:len_gcd]
        if all(str1[i:i+len_gcd]==c for i in range(0,len1,len_gcd)) and\
        all(str2[i:i+len_gcd]==c for i in range(0,len2,len_gcd)):
         return c
        return ""
    def gcd(self , a , b):
        return a if b == 0 else self.gcd(b,a%b)

File: gcd_string_high.cpp
class Solution {
public:
    string gcdOfStrings(string str1, string str2) {
        int len1 = str1.length();
        int len2 = str2.length();
        int len = gdc(len1,len2);
        if(len1%len!=0||len2%len!=0) return "";
        string c = str1.substr(0,len);
        for(int i = 0;i<len1;i = i+len){
            if(str1.substr(i,len)!=c) return "";
        }
        for(int j = 0;j<len2;j = j+len){
            if(str2.substr(j,len)!=c) return "";
        }
        return c;
    }
public:       
     int gdc(int a,int  b){
        return b == 0? a:gdc(b,a%b);
        }
};
